Fix wxFontList::FindOrCreateFont() for pixel and fractional point sizes

PBfordev opened the pull request:
This is an attempt to fix #23139. From the two approaches suggested by vadz in the referenced issue, I have chosen "just make it work" one instead of "make it work and pretty". The upside is that it can be easily backported. Yes that was the reason for my choice. ;P

Regarding the added tests. Instead of adding a new test file for `wxFontList` (which would require rebaking), I added one `TEST_CASE` to existing `fonttest.cpp`.

The new tests are very simple (but IMO sufficient); however, I am not sure if any tests in `fonttest.cpp` are actually run and from which test executable (gui or drawing?). I am also not sure about comparing doubles with `==`  and `Approx()` should probably have been used there.

PBfordev committed:
Fix wxFontList::FindOrCreateFont() for pixel and fractional point sizes

wxFontList::FindOrCreateFont() worked properly only for integral point sizes.

Make it work also when a pixel size or fractional point size is used.

PBfordev commented:
I don't understand the test failures.

Even on Windows it fails: when I request point size 10.5, I get 11: e.g., [here](https://github.com/wxWidgets/wxWidgets/actions/runs/3914869115/jobs/6692417525#step:10:24). However, I get 10.5 on my Windows 10 computer as expected? Are default fonts on CIs not supporting fractional sizes?

I will not attempt to make further changes to the PR and will wait for suggestions on how to continue with this.

vadz committed:
Fix fractional point size comparison in the test

vadz commented:
Err, I really don't understand how can it pass for you, `GetPointSize()` returns an int which can never be equal to `10.5`. Replacing it with `GetFractionalPointSize()` should make it work, so I've pushed a commit doing this to check...

PBfordev commented:
> Err, I really don't understand how can it pass for you

Sorry, it did not. I tested in a separate code where I actually did use `GetFractionalPointSize()`.

vadz commented:
OK, this would explain it :-)

The failures in Unix tests are also understandable and I actually wonder if wxMSW shouldn't behave like this too: it makes sense to specify 0 width when creating the font, but not when querying it for size. In any case, we could just compare `size.y` here.

But there is also a failure under Mac which is more worrisome:
```
./font/fonttest.cpp:505: FAILED:
  REQUIRE( font1->GetPixelSize() == pixelSize )
with expansion:
  16*31 == 0*32
```
This one might be a real bug...

Will you be able to look at it or should I try to?

vadz committed:
Avoid duplication in the test by using SECTIONs

No real changes.

vadz committed:
Show font properties if any checks fail

This will make it simpler to debug any failures in the test, especially
in the CI builds where debugger can't be used.

vadz committed:
Limit comparison of pixel sizes to the height only

In wxMSW wxFont::GetPixelSize() returns the size that the font was
originally created from while in the other ports it returns the actual
size being used.

wxMSW behaviour is probably wrong and should be changed, but for now
just restrict the comparison to the vertical component to avoid the test
failures.

vadz committed:
Slightly reformat FindOrCreateFont() declaration and add comments

Explain that the new overload is preferred and why.

vadz committed:
Round the value returned from wxFont::GetPixelSize() in wxOSX

It seems wrong to return height of 12 from this function when
GetTextExtent() returns 12.9, so round the returned values instead of
truncating them to int.

This makes wxOSX more consistent with the other ports and fixes one of
the failures in the recently added wxFontList unit test.

vadz committed:
Avoid calling wxFont::GetPixelSize() under wxOSX, it's buggy

Work around a bug in wxOSX which breaks the test if we call DumpFont()
because this function calls wxFont::GetPixelSize() which actually
changes the size of the font if it was specified in pixels.

This should be reverted once the bug is fixed.

See #23144.

vadz committed:
Fix using pixel size with width 0 in FindOrCreateFont()

This width value is special and means that we don't care about the
width, so only compare the heights in this case.

vadz committed:
Update workaround for handling wxFONTFAMILY_DEFAULT in wxOSX

The existing workaround was not enough because wxFont::GetFamily() can
return _both_ wxFONTFAMILY_DEFAULT and wxFONTFAMILY_SWISS for the fonts
using the default family under wxOSX.

This is, of course, wrong and ideally this workaround should be removed
completely once #23144 is fixed, but for now this at least makes
wxFontList work correctly for the fonts using sizes in pixels under
macOS too.

vadz committed:
Avoid an extraneous font copy in FindOrCreateFont()

There doesn't seem to be any point in creating the font and then
allocating a copy of it on the heap instead of creating the font on the
heap directly, so just do this instead.

No real changes, this is just a micro-optimization.

vadz committed:
Improve FindOrCreateFont() documentation

Explain that the overload taking wxFontInfo is preferred and why,
mention that the returned pointer must not be deleted and also give a
trivial example of using this function.

vadz commented:
I really shouldn't have spent half of my Saturday on this but, well, I did... unfortunately the situation is pretty bad under Mac and I can't fix all the problems there, so we'll have to just work around for them for now (as it was already the case before...), but at least we should be able to merge this if the tests pass now.

PBfordev commented:
> Will you be able to look at it or should I try to?

Sorry but no, I am Windows only.

vadz commented:
If the latest version works for you, let's merge it. Please let me know if you have any objections.

PBfordev commented:
It works. Thanks for the quick fix and sorry for ruining your Saturday.

vadz commented:
Thanks for testing! And no, it wasn't ruined at all, it's just that I hoped to do something probably more important in wx that I won't have time to do any longer today, but it's solely my fault for being so easily distracted. And the fact that this allowed us to find and already partially fix #23144 means that this was more important than I had thought, too, so maybe it's all for the best.

Anyhow, I'll merge this soon.
